pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/stats/MapColumnPreIndexStatsCollector.java

```diff
@@ -0,0 +1,288 @@
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.pinot.common.utils.PinotDataType;
+import org.apache.pinot.segment.spi.creator.StatsCollectorConfig;
+import org.apache.pinot.spi.config.table.TableConfig;
+import org.apache.pinot.spi.config.table.TableType;
+import org.apache.pinot.spi.data.DimensionFieldSpec;
+import org.apache.pinot.spi.data.FieldSpec;
+import org.apache.pinot.spi.data.Schema;
+import org.apache.pinot.spi.utils.MapUtils;
+import org.apache.pinot.spi.utils.builder.TableConfigBuilder;
+
+
+/**
+ * Extension of {@link AbstractColumnStatisticsCollector} for Map column type.
+ *
+ * The Map column type is different than other columns in that it is essentially recursive.  It contains keys
+ * and those keys are analogous to columns and, as such, have Key level statistics. So, this class keeps track of
+ * Map column level statistics _and_ Key level statistics.  The Key Level statistics can then be used during
+ * the creation of the Immutable Segment to make decisions about how keys will be stored or what Map data structure
+ * to use.
+ *
+ * Assumptions that are made:
+ * 1. Each key has a single type for the value's associated with it across all documents.
+ * 2. At this point in the  Pinot process, the type consistency of a key should already be enforced, so if a
+ * heterogenous value types for a key are encountered will constructing the Map statistics it can be raised as a
+ * fault.
+ */
+public class MapColumnPreIndexStatsCollector extends AbstractColumnStatisticsCollector {
+  private ObjectOpenHashSet<String> _keys = new ObjectOpenHashSet<>(INITIAL_HASH_SET_SIZE);
+  private final HashMap<String, AbstractColumnStatisticsCollector> _keyStats;
+  private String[] _sortedValues;
+  private int _minLength = Integer.MAX_VALUE;
+  private int _maxLength = 0;
+  private int _maxRowLength = 0;
+  private String _minValue = null;
+  private String _maxValue = null;
+  private boolean _sealed = false;
+  private final String _column;
+
+  public MapColumnPreIndexStatsCollector(String column, StatsCollectorConfig statsCollectorConfig) {
+    super(column, statsCollectorConfig);
+    super._sorted = false;
+    _keyStats = new HashMap<>();
+    _column = column;
+  }
+
+  public AbstractColumnStatisticsCollector getKeyStatistics(String key) {
+    return _keyStats.get(key);
+  }
+
+  @Override
+  public void collect(Object entry) {
+    assert !_sealed;
+
+    if (entry instanceof Map) {
+      final Map<String, Object> mapValue = (Map<String, Object>) entry;
+      byte[] serializeMap = MapUtils.serializeMap(mapValue);
+
+      _maxRowLength = Math.max(_maxRowLength, serializeMap.length);
+      for (Map.Entry<String, Object> mapValueEntry : mapValue.entrySet()) {
+        final String key = mapValueEntry.getKey();
+
+        // Record statistics about the key
+        int length = serializeMap.length;
+        if (_keys.add(key)) {
+          if (isPartitionEnabled()) {
```

Comment:
Can the table be partitioned on Map keys?

Reply:
Yes. 
However if really the partition key is a map key, then it's better to extract map key to a new column for efficiency.